// ng-indexed-db/esm2015/lib/indexed-db.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IndexedDB, NG_DATABASES } from './core/indexed-db.service';
import { NgDatabaseFactory } from './core/factory/database-factory';
import { IdGenerator } from './core/id-generator/id-generator';
export class IndexedDBModule {
    /**
     * @param {?=} databases
     * @return {?}
     */
    static forRoot(databases = []) {
        return {
            ngModule: IndexedDBModule,
            providers: [
                { provide: NG_DATABASES, useValue: databases }
            ]
        };
    }
}
IndexedDBModule.decorators = [
    { type: NgModule, args: [{
                declarations: [],
                imports: [CommonModule],
                exports: [],
                providers: [
                    IndexedDB,
                    NgDatabaseFactory,
                    IdGenerator,
                ],
            },] }
];
