// @progress/kendo-angular-scheduler/dist/npm/editing/recurrence/recurrence-end-rule-editor.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var recurrence_service_1 = require("./recurrence.service");
var util_1 = require("../../common/util");
var kendo_angular_l10n_1 = require("@progress/kendo-angular-l10n");
var end_rule_radio_button_directive_1 = require("./end-rule-radio-button.directive");
var kendo_date_math_1 = require("@progress/kendo-date-math");
var utils_1 = require("../../views/utils");
/**
 * @hidden
 */
var RecurrenceEndRuleEditorComponent = /** @class */ (function () {
    function RecurrenceEndRuleEditorComponent(recurrence, localization) {
        this.recurrence = recurrence;
        this.localization = localization;
        this.numericOptions = {
            min: 1,
            format: '#',
            autoCorrect: true,
            step: 1,
            spinners: true
        };
        this.datePickerOptions = {
            activeView: 'month',
            bottomView: 'month',
            topView: 'century',
            disabledDatesValidation: true,
            navigation: true,
            format: 'd'
        };
        this.uniqueId = this.recurrence.getUniqueId();
        this.setInitialValues();
        this.subscribeChanges();
    }
    Object.defineProperty(RecurrenceEndRuleEditorComponent.prototype, "userNumericOptions", {
        set: function (options) {
            this.numericOptions = tslib_1.__assign({}, this.numericOptions, options);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RecurrenceEndRuleEditorComponent.prototype, "userDatePickerOptions", {
        set: function (options) {
            this.datePickerOptions = tslib_1.__assign({}, this.datePickerOptions, options);
        },
        enumerable: true,
        configurable: true
    });
    RecurrenceEndRuleEditorComponent.prototype.ngOnDestroy = function () {
        if (this.subscriptions) {
            this.subscriptions.unsubscribe();
        }
    };
    RecurrenceEndRuleEditorComponent.prototype.setEndRule = function (endRule) {
        if (endRule === 'count') {
            this.recurrence.rrule.count = this.countValue;
        }
        else if (endRule === 'until') {
            this.recurrence.until = this.untilValue;
        }
    };
    Object.defineProperty(RecurrenceEndRuleEditorComponent.prototype, "rrule", {
        get: function () {
            return this.recurrence.rrule;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RecurrenceEndRuleEditorComponent.prototype, "isCountDisabled", {
        get: function () {
            return this.recurrence.endRule !== 'count';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RecurrenceEndRuleEditorComponent.prototype, "isUntilDisabled", {
        get: function () {
            return this.recurrence.endRule !== 'until';
        },
        enumerable: true,
        configurable: true
    });
    RecurrenceEndRuleEditorComponent.prototype.onCountChange = function (value) {
        if (util_1.isPresent(value)) {
            this.recurrence.count = value;
        }
    };
    RecurrenceEndRuleEditorComponent.prototype.onCountBlur = function () {
        if (!util_1.isPresent(this.countValue)) {
            this.recurrence.count = this.countValue = 1;
        }
    };
    RecurrenceEndRuleEditorComponent.prototype.onUntilChange = function (value) {
        if (util_1.isPresent(value)) {
            this.recurrence.until = this.createUntil(value);
        }
    };
    RecurrenceEndRuleEditorComponent.prototype.onUntilBlur = function () {
        if (!util_1.isPresent(this.untilValue)) {
            this.recurrence.until = this.untilValue = this.createUntil(this.recurrence.start);
        }
    };
    RecurrenceEndRuleEditorComponent.prototype.textFor = function (key) {
        return this.localization.get(key);
    };
    RecurrenceEndRuleEditorComponent.prototype.onEndLabelClick = function () {
        var selected = this.endRuleRadioButtons.toArray().find(function (r) { return r.elem.checked; });
        if (selected) {
            selected.elem.focus();
        }
    };
    RecurrenceEndRuleEditorComponent.prototype.setInitialValues = function () {
        this.countValue = this.rrule.count || 1;
        var currentUntil = this.recurrence.until;
        var currentStart = this.recurrence.start;
        this.untilValue = util_1.isPresent(currentUntil) ? currentUntil : this.createUntil(currentStart);
    };
    RecurrenceEndRuleEditorComponent.prototype.subscribeChanges = function () {
        var _this = this;
        this.subscriptions = this.recurrence.endRuleChange.subscribe(function (endRule) {
            _this.setEndRule(endRule);
        });
        this.subscriptions.add(this.recurrence.frequencyChange.subscribe(function () {
            _this.setInitialValues();
        }));
    };
    RecurrenceEndRuleEditorComponent.prototype.createUntil = function (until) {
        // Read the until date as UTC date parts to avoid interfering with the local time zone.
        var untilDate = utils_1.toUTCDate(until);
        untilDate.setUTCDate(untilDate.getUTCDate() + 1);
        // Convert to the scheduler time zone.
        return kendo_date_math_1.ZonedDate.fromUTCDate(untilDate, this.recurrence.timezone).toLocalDate();
    };
    RecurrenceEndRuleEditorComponent.decorators = [
        { type: core_1.Component, args: [{
                    selector: 'kendo-recurrence-end-rule-editor',
                    template: "\n        <div class='k-edit-label'>\n            <label (click)=\"onEndLabelClick()\">{{ textFor('endLabel') }}</label>\n        </div>\n        <div class='k-edit-field'>\n            <ul class='k-reset'>\n                <li>\n                    <input kendoRecurrenceEndRuleRadioButton='k-endrule-never-{{uniqueId}}' />\n                    <label class='k-radio-label' for='k-endrule-never-{{uniqueId}}'>{{ textFor('endNever') }}</label>\n                </li>\n                <li>\n                    <input kendoRecurrenceEndRuleRadioButton='k-endrule-count-{{uniqueId}}' />\n                    <label class='k-radio-label' for='k-endrule-count-{{uniqueId}}'>{{ textFor('endAfter') }}</label>\n\n                    <kendo-numerictextbox\n                        [style.width.px]='70'\n                        [autoCorrect]='numericOptions.autoCorrect'\n                        [decimals]='0'\n                        [disabled]='isCountDisabled'\n                        [format]=\"numericOptions.format\"\n                        [min]='numericOptions.min'\n                        [max]=\"numericOptions.max\"\n                        [readonly]=\"numericOptions.readonly\"\n                        [selectOnFocus]=\"numericOptions.selectOnFocus\"\n                        [spinners]=\"numericOptions.spinners\"\n                        [step]=\"numericOptions.step\"\n                        [title]=\"numericOptions.title\"\n                        [(value)]='countValue'\n                        (blur)=\"onCountBlur()\"\n                        (valueChange)='onCountChange($event)'>\n                    </kendo-numerictextbox>\n                    <span>{{ textFor('endOccurrence') }}</span>\n                </li>\n                <li>\n                    <input kendoRecurrenceEndRuleRadioButton='k-endrule-until-{{uniqueId}}' />\n                    <label class='k-radio-label' for='k-endrule-until-{{uniqueId}}'>{{ textFor('endOn') }}</label>\n\n                    <kendo-datepicker\n                        [style.width.px]='150'\n                        [disabled]='isUntilDisabled'\n                        [activeView]=\"datePickerOptions.activeView\"\n                        [bottomView]=\"datePickerOptions.bottomView\"\n                        [disabledDatesValidation]=\"datePickerOptions.disabledDatesValidation\"\n                        [focusedDate]=\"datePickerOptions.focusedDate\"\n                        [format]=\"datePickerOptions.format\"\n                        [formatPlaceholder]=\"datePickerOptions.formatPlaceHolder\"\n                        [max]=\"datePickerOptions.max\"\n                        [min]=\"datePickerOptions.min\"\n                        [navigation]=\"datePickerOptions.navigation\"\n                        [placeholder]=\"datePickerOptions.placeholder\"\n                        [readOnlyInput]=\"datePickerOptions.readOnlyInput\"\n                        [readonly]=\"datePickerOptions.readonly\"\n                        [title]=\"datePickerOptions.title\"\n                        [topView]=\"datePickerOptions.topView\"\n                        [weekNumber]=\"datePickerOptions.weekNumber\"\n                        [disabledDates]=\"datePickerOptions.disabledDates\"\n                        [popupSettings]=\"datePickerOptions.popupSettings\"\n                        [(value)]='untilValue'\n                        (blur)=\"onUntilBlur()\"\n                        (valueChange)='onUntilChange($event)'>\n                    </kendo-datepicker>\n                </li>\n            </ul>\n        </div>\n    "
                },] },
    ];
    /** @nocollapse */
    RecurrenceEndRuleEditorComponent.ctorParameters = function () { return [
        { type: recurrence_service_1.RecurrenceService },
        { type: kendo_angular_l10n_1.LocalizationService }
    ]; };
    RecurrenceEndRuleEditorComponent.propDecorators = {
        userNumericOptions: [{ type: core_1.Input }],
        userDatePickerOptions: [{ type: core_1.Input }],
        endRuleRadioButtons: [{ type: core_1.ViewChildren, args: [end_rule_radio_button_directive_1.EndRuleRadioButtonDirective,] }]
    };
    return RecurrenceEndRuleEditorComponent;
}());
exports.RecurrenceEndRuleEditorComponent = RecurrenceEndRuleEditorComponent;
