// ng-indexed-db/esm5/lib/core/indexed-db.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, InjectionToken, Inject } from '@angular/core';
import { switchMap, skipWhile } from 'rxjs/operators';
import { NgDatabaseFactory } from './factory/database-factory';
import { IdGenerator } from './id-generator/id-generator';
/**
 * Injection token that gives the databases metadata
 * @type {?}
 */
export var NG_DATABASES = new InjectionToken('NG_DATABASES');
/**
 * Service that handles all Indexed Databases and
 * tables requests.
 */
var IndexedDB = /** @class */ (function () {
    function IndexedDB(databases, idGenerator, databaseFactory) {
        this.databases = databases;
        this.idGenerator = idGenerator;
        this.databaseFactory = databaseFactory;
        this.databaseFactory.createConnnections(this.databases);
    }
    Object.defineProperty(IndexedDB.prototype, "onReady", {
        /**
         * Indicates if all databases have been created.
         * Only emits an event when recives a true value.
         */
        get: /**
         * Indicates if all databases have been created.
         * Only emits an event when recives a true value.
         * @return {?}
         */
        function () {
            return this.databaseFactory.connectionsReady
                .pipe(skipWhile((/**
             * @param {?} event
             * @return {?}
             */
            function (event) { return !event; })));
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Request a list with all elements
     * @param storeName Store name
     * @param databaseName Database name
     */
    /**
     * Request a list with all elements
     * @template M
     * @param {?} storeName Store name
     * @param {?=} databaseName Database name
     * @return {?}
     */
    IndexedDB.prototype.list = /**
     * Request a list with all elements
     * @template M
     * @param {?} storeName Store name
     * @param {?=} databaseName Database name
     * @return {?}
     */
    function (storeName, databaseName) {
        var _this = this;
        return this.onReady.pipe(switchMap((/**
         * @return {?}
         */
        function () { return _this.databaseFactory.getDatabase(databaseName).list(storeName); })));
    };
    /**
     * Request an element by key
     * @param storeName Store name
     * @param key Key value to find the element
     * @param databaseName Database name
     */
    /**
     * Request an element by key
     * @template M
     * @param {?} storeName Store name
     * @param {?} key Key value to find the element
     * @param {?=} databaseName Database name
     * @return {?}
     */
    IndexedDB.prototype.get = /**
     * Request an element by key
     * @template M
     * @param {?} storeName Store name
     * @param {?} key Key value to find the element
     * @param {?=} databaseName Database name
     * @return {?}
     */
    function (storeName, key, databaseName) {
        var _this = this;
        return this.onReady.pipe(switchMap((/**
         * @return {?}
         */
        function () { return _this.databaseFactory.getDatabase(databaseName).get(storeName, key); })));
    };
    /**
     * Request to store an element
     * @param storeName Store name
     * @param databaseName Database name
     */
    /**
     * Request to store an element
     * @template M
     * @param {?} storeName Store name
     * @param {?} data
     * @param {?=} databaseName Database name
     * @return {?}
     */
    IndexedDB.prototype.create = /**
     * Request to store an element
     * @template M
     * @param {?} storeName Store name
     * @param {?} data
     * @param {?=} databaseName Database name
     * @return {?}
     */
    function (storeName, data, databaseName) {
        var _this = this;
        // tslint:disable-next-line: no-string-literal
        if (!data['id']) {
            data['id'] = this.idGenerator.generate();
        }
        return this.onReady.pipe(switchMap((/**
         * @return {?}
         */
        function () { return _this.databaseFactory.getDatabase(databaseName).create(storeName, data); })));
    };
    /**
     * Request to update a stored element
     * @param storeName Store name
     * @param databaseName Database name
     */
    /**
     * Request to update a stored element
     * @template M
     * @param {?} storeName Store name
     * @param {?} data
     * @param {?=} databaseName Database name
     * @return {?}
     */
    IndexedDB.prototype.update = /**
     * Request to update a stored element
     * @template M
     * @param {?} storeName Store name
     * @param {?} data
     * @param {?=} databaseName Database name
     * @return {?}
     */
    function (storeName, data, databaseName) {
        var _this = this;
        return this.onReady.pipe(switchMap((/**
         * @return {?}
         */
        function () { return _this.databaseFactory.getDatabase(databaseName).update(storeName, data); })));
    };
    /**
     * Request to delete a stored element
     * @param storeName Store name
     * @param databaseName Database name
     */
    /**
     * Request to delete a stored element
     * @param {?} storeName Store name
     * @param {?} key
     * @param {?=} databaseName Database name
     * @return {?}
     */
    IndexedDB.prototype.delete = /**
     * Request to delete a stored element
     * @param {?} storeName Store name
     * @param {?} key
     * @param {?=} databaseName Database name
     * @return {?}
     */
    function (storeName, key, databaseName) {
        var _this = this;
        return this.onReady.pipe(switchMap((/**
         * @return {?}
         */
        function () { return _this.databaseFactory.getDatabase(databaseName).delete(storeName, key); })));
    };
    IndexedDB.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    IndexedDB.ctorParameters = function () { return [
        { type: Array, decorators: [{ type: Inject, args: [NG_DATABASES,] }] },
        { type: IdGenerator },
        { type: NgDatabaseFactory }
    ]; };
    return IndexedDB;
}());
export { IndexedDB };
if (false) {
    /**
     * @type {?}
     * @private
     */
    IndexedDB.prototype.databases;
    /**
     * @type {?}
     * @private
     */
    IndexedDB.prototype.idGenerator;
    /**
     * @type {?}
     * @private
     */
    IndexedDB.prototype.databaseFactory;
}
