// @progress/kendo-angular-scheduler/dist/es2015/views/common/hint-container.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Component, ChangeDetectorRef, ContentChild, TemplateRef } from '@angular/core';
/**
 * @hidden
 */
export class HintContainerComponent {
    constructor(changeDetector) {
        this.changeDetector = changeDetector;
        this.changeDetector.detach();
    }
    detectChanges() {
        this.changeDetector.detectChanges();
    }
}
HintContainerComponent.decorators = [
    { type: Component, args: [{
                selector: 'kendo-hint-container',
                template: `
        <ng-container [ngTemplateOutlet]="hintTemplate">
        </ng-container>
    `
            },] },
];
/** @nocollapse */
HintContainerComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
HintContainerComponent.propDecorators = {
    hintTemplate: [{ type: ContentChild, args: [TemplateRef,] }]
};
