// @progress/kendo-angular-treeview/dist/es2015/drag-and-drop/drop-hint/drop-hint.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Component, HostBinding, ChangeDetectionStrategy } from "@angular/core";
/**
 * @hidden
 */
export class DropHintComponent {
    constructor() {
        this.hostClass = true;
        this.position = 'fixed';
        this.pointerEvents = 'none';
    }
}
DropHintComponent.decorators = [
    { type: Component, args: [{
                changeDetection: ChangeDetectionStrategy.OnPush,
                selector: 'kendo-treeview-drop-hint',
                template: `
        <div
            *ngIf="!template"
            class="k-drop-hint k-drop-hint-h"
        >
            <div class='k-drop-hint-start'></div>
            <div class='k-drop-hint-line'></div>
        </div>

        <ng-template
            *ngIf="template"
            [ngTemplateOutlet]="template"
        >
        <ng-template>
    `
            },] },
];
DropHintComponent.propDecorators = {
    hostClass: [{ type: HostBinding, args: ['class.k-drop-hint-container',] }],
    position: [{ type: HostBinding, args: ['style.position',] }],
    pointerEvents: [{ type: HostBinding, args: ['style.pointer-events',] }]
};
