// @progress/kendo-angular-treeview/dist/es/drag-and-drop/drop-hint/drop-hint.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Component, HostBinding, ChangeDetectionStrategy } from "@angular/core";
/**
 * @hidden
 */
var DropHintComponent = /** @class */ (function () {
    function DropHintComponent() {
        this.hostClass = true;
        this.position = 'fixed';
        this.pointerEvents = 'none';
    }
    DropHintComponent.decorators = [
        { type: Component, args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-treeview-drop-hint',
                    template: "\n        <div\n            *ngIf=\"!template\"\n            class=\"k-drop-hint k-drop-hint-h\"\n        >\n            <div class='k-drop-hint-start'></div>\n            <div class='k-drop-hint-line'></div>\n        </div>\n\n        <ng-template\n            *ngIf=\"template\"\n            [ngTemplateOutlet]=\"template\"\n        >\n        <ng-template>\n    "
                },] },
    ];
    DropHintComponent.propDecorators = {
        hostClass: [{ type: HostBinding, args: ['class.k-drop-hint-container',] }],
        position: [{ type: HostBinding, args: ['style.position',] }],
        pointerEvents: [{ type: HostBinding, args: ['style.pointer-events',] }]
    };
    return DropHintComponent;
}());
export { DropHintComponent };
