// @progress/kendo-angular-treeview/dist/es2015/flat-binding.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, Input } from '@angular/core';
import { isChanged } from '@progress/kendo-angular-common';
import { getter } from '@progress/kendo-common';
import { of } from 'rxjs';
import { compose } from './funcs';
import { TreeViewComponent } from './treeview.component';
import { isBlank, isNullOrEmptyString, isPresent } from './utils';
import { FlatEditingService } from './drag-and-drop/editing-services/flat-editing.service';
const findChildren = (prop, nodes, value) => nodes.filter(x => prop(x) === value);
const ɵ0 = findChildren;
/**
 * A directive which encapsulates the retrieval of the child nodes.
 */
export class FlatDataBindingDirective {
    constructor(treeView) {
        this.treeView = treeView;
        /**
         * @hidden
         */
        this.originalData = [];
    }
    /**
     * The nodes which will be displayed by the TreeView.
     */
    set nodes(values) {
        this.originalData = values || [];
        if (!isNullOrEmptyString(this.parentIdField)) {
            const prop = getter(this.parentIdField);
            this.treeView.nodes = (this.originalData).filter(compose(isBlank, prop));
        }
        else {
            this.treeView.nodes = this.originalData.slice(0);
        }
    }
    /**
     * @hidden
     */
    ngOnInit() {
        if (isPresent(this.parentIdField) && isPresent(this.idField)) {
            const fetchChildren = node => findChildren(getter(this.parentIdField), this.originalData || [], getter(this.idField)(node));
            this.treeView.hasChildren = node => fetchChildren(node).length > 0;
            this.treeView.children = node => of(fetchChildren(node));
            this.treeView.editService = new FlatEditingService(this);
        }
    }
    /**
     * @hidden
     */
    ngOnChanges(changes) {
        if (isChanged("parentIdField", changes, false)) {
            this.nodes = this.originalData;
        }
    }
}
FlatDataBindingDirective.decorators = [
    { type: Directive, args: [{ selector: '[kendoTreeViewFlatDataBinding]' },] },
];
/** @nocollapse */
FlatDataBindingDirective.ctorParameters = () => [
    { type: TreeViewComponent }
];
FlatDataBindingDirective.propDecorators = {
    nodes: [{ type: Input }],
    parentIdField: [{ type: Input }],
    idField: [{ type: Input }]
};
export { ɵ0 };
