// @datorama/akita-ngdevtools/esm5/lib/ng-devtools.module.js
import * as tslib_1 from "tslib";
import { APP_INITIALIZER, ModuleWithProviders, NgModule, NgZone, InjectionToken, Inject, Injectable } from '@angular/core';
import { akitaDevtools } from '@datorama/akita';
import * as i0 from "@angular/core";
export var DEVTOOLS_OPTIONS = new InjectionToken('DevtoolsOptions');
var AkitaDevtools = /** @class */ (function () {
    function AkitaDevtools(ngZone, options) {
        this.ngZone = ngZone;
        this.options = options;
        akitaDevtools(this.ngZone, this.options);
    }
    AkitaDevtools.ctorParameters = function () { return [
        { type: NgZone },
        { type: undefined, decorators: [{ type: Inject, args: [DEVTOOLS_OPTIONS,] }] }
    ]; };
    AkitaDevtools.ɵprov = i0.ɵɵdefineInjectable({ factory: function AkitaDevtools_Factory() { return new AkitaDevtools(i0.ɵɵinject(i0.NgZone), i0.ɵɵinject(DEVTOOLS_OPTIONS)); }, token: AkitaDevtools, providedIn: "root" });
    AkitaDevtools = tslib_1.__decorate([
        Injectable({
            providedIn: 'root'
        }),
        tslib_1.__param(1, Inject(DEVTOOLS_OPTIONS)),
        tslib_1.__metadata("design:paramtypes", [NgZone, Object])
    ], AkitaDevtools);
    return AkitaDevtools;
}());
export { AkitaDevtools };
// auto initialize the devtools
export function d() { }
export function init(akitaDevtools) {
    return d;
}
var AkitaNgDevtools = /** @class */ (function () {
    function AkitaNgDevtools() {
    }
    AkitaNgDevtools_1 = AkitaNgDevtools;
    AkitaNgDevtools.forRoot = function (options) {
        return {
            ngModule: AkitaNgDevtools_1,
            providers: [
                {
                    provide: DEVTOOLS_OPTIONS,
                    useValue: options
                },
                {
                    provide: APP_INITIALIZER,
                    useFactory: init,
                    deps: [AkitaDevtools],
                    multi: true
                }
            ]
        };
    };
    var AkitaNgDevtools_1;
    AkitaNgDevtools = AkitaNgDevtools_1 = tslib_1.__decorate([
        NgModule({})
    ], AkitaNgDevtools);
    return AkitaNgDevtools;
}());
export { AkitaNgDevtools };
