// @datorama/akita/esm5/lib/actions.js
export var currentAction = {
    type: null,
    entityIds: null,
    skip: false,
};
var customActionActive = false;
export function resetCustomAction() {
    customActionActive = false;
}
// public API for custom actions. Custom action always wins
export function logAction(type, entityIds) {
    setAction(type, entityIds);
    customActionActive = true;
}
export function setAction(type, entityIds) {
    if (customActionActive === false) {
        currentAction.type = type;
        currentAction.entityIds = entityIds;
    }
}
export function setSkipAction(skip) {
    if (skip === void 0) { skip = true; }
    currentAction.skip = skip;
}
export function action(action, entityIds) {
    return function (target, propertyKey, descriptor) {
        var originalMethod = descriptor.value;
        descriptor.value = function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            logAction(action, entityIds);
            return originalMethod.apply(this, args);
        };
        return descriptor;
    };
}
