// ng-indexed-db/esm2015/lib/core/indexed-db.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, InjectionToken, Inject } from '@angular/core';
import { switchMap, skipWhile } from 'rxjs/operators';
import { NgDatabaseFactory } from './factory/database-factory';
import { IdGenerator } from './id-generator/id-generator';
/**
 * Injection token that gives the databases metadata
 * @type {?}
 */
export const NG_DATABASES = new InjectionToken('NG_DATABASES');
/**
 * Service that handles all Indexed Databases and
 * tables requests.
 */
export class IndexedDB {
    /**
     * @param {?} databases
     * @param {?} idGenerator
     * @param {?} databaseFactory
     */
    constructor(databases, idGenerator, databaseFactory) {
        this.databases = databases;
        this.idGenerator = idGenerator;
        this.databaseFactory = databaseFactory;
        this.databaseFactory.createConnnections(this.databases);
    }
    /**
     * Indicates if all databases have been created.
     * Only emits an event when recives a true value.
     * @return {?}
     */
    get onReady() {
        return this.databaseFactory.connectionsReady
            .pipe(skipWhile((/**
         * @param {?} event
         * @return {?}
         */
        (event) => !event)));
    }
    /**
     * Request a list with all elements
     * @template M
     * @param {?} storeName Store name
     * @param {?=} databaseName Database name
     * @return {?}
     */
    list(storeName, databaseName) {
        return this.onReady.pipe(switchMap((/**
         * @return {?}
         */
        () => this.databaseFactory.getDatabase(databaseName).list(storeName))));
    }
    /**
     * Request an element by key
     * @template M
     * @param {?} storeName Store name
     * @param {?} key Key value to find the element
     * @param {?=} databaseName Database name
     * @return {?}
     */
    get(storeName, key, databaseName) {
        return this.onReady.pipe(switchMap((/**
         * @return {?}
         */
        () => this.databaseFactory.getDatabase(databaseName).get(storeName, key))));
    }
    /**
     * Request to store an element
     * @template M
     * @param {?} storeName Store name
     * @param {?} data
     * @param {?=} databaseName Database name
     * @return {?}
     */
    create(storeName, data, databaseName) {
        // tslint:disable-next-line: no-string-literal
        if (!data['id']) {
            data['id'] = this.idGenerator.generate();
        }
        return this.onReady.pipe(switchMap((/**
         * @return {?}
         */
        () => this.databaseFactory.getDatabase(databaseName).create(storeName, data))));
    }
    /**
     * Request to update a stored element
     * @template M
     * @param {?} storeName Store name
     * @param {?} data
     * @param {?=} databaseName Database name
     * @return {?}
     */
    update(storeName, data, databaseName) {
        return this.onReady.pipe(switchMap((/**
         * @return {?}
         */
        () => this.databaseFactory.getDatabase(databaseName).update(storeName, data))));
    }
    /**
     * Request to delete a stored element
     * @param {?} storeName Store name
     * @param {?} key
     * @param {?=} databaseName Database name
     * @return {?}
     */
    delete(storeName, key, databaseName) {
        return this.onReady.pipe(switchMap((/**
         * @return {?}
         */
        () => this.databaseFactory.getDatabase(databaseName).delete(storeName, key))));
    }
}
IndexedDB.decorators = [
    { type: Injectable }
];
/** @nocollapse */
IndexedDB.ctorParameters = () => [
    { type: Array, decorators: [{ type: Inject, args: [NG_DATABASES,] }] },
    { type: IdGenerator },
    { type: NgDatabaseFactory }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    IndexedDB.prototype.databases;
    /**
     * @type {?}
     * @private
     */
    IndexedDB.prototype.idGenerator;
    /**
     * @type {?}
     * @private
     */
    IndexedDB.prototype.databaseFactory;
}
