// ng-indexed-db/esm5/lib/indexed-db.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IndexedDB, NG_DATABASES } from './core/indexed-db.service';
import { NgDatabaseFactory } from './core/factory/database-factory';
import { IdGenerator } from './core/id-generator/id-generator';
var IndexedDBModule = /** @class */ (function () {
    function IndexedDBModule() {
    }
    /**
     * @param {?=} databases
     * @return {?}
     */
    IndexedDBModule.forRoot = /**
     * @param {?=} databases
     * @return {?}
     */
    function (databases) {
        if (databases === void 0) { databases = []; }
        return {
            ngModule: IndexedDBModule,
            providers: [
                { provide: NG_DATABASES, useValue: databases }
            ]
        };
    };
    IndexedDBModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [],
                    imports: [CommonModule],
                    exports: [],
                    providers: [
                        IndexedDB,
                        NgDatabaseFactory,
                        IdGenerator,
                    ],
                },] }
    ];
    return IndexedDBModule;
}());
export { IndexedDBModule };
