// @progress/kendo-angular-treelist/dist/es2015/navigation/focusable.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, ElementRef, Input, Optional, Renderer2 } from '@angular/core';
import { DefaultFocusableElement } from './default-focusable-element';
import { FocusGroup } from './focus-group';
/**
 * A directive that controls the way focusable cell elements receive
 * [focus in a navigable TreeList]({% slug keyboard_navigation_treelist %}).
 *
 * The directive takes as input the `cellContext` from the cell template.
 *
 * ```ts-no-run
 * <kendo-treelist-column>
 *   <ng-template kendoTreeListCellTemplate let-cellContext="cellContext">
 *     <input type="button" class="k-button" [kendoTreeListFocusable]="cellContext" value="Focus me!" />
 *   </ng-template>
 * </kendo-treelist-column>
 * ```
 */
export class FocusableDirective {
    constructor(hostElement, renderer, group) {
        this.hostElement = hostElement;
        this.renderer = renderer;
        this.group = group;
        this.active = true;
    }
    /**
     * @hidden
     */
    set addCellContext(value) { this.cellContext = value; }
    /**
     * @hidden
     */
    set editCellContext(value) { this.cellContext = value; }
    /**
     * @hidden
     */
    set removeCellContext(value) { this.cellContext = value; }
    /**
     * @hidden
     */
    set saveCellContext(value) { this.cellContext = value; }
    /**
     * @hidden
     */
    set cancelCellContext(value) { this.cellContext = value; }
    ngOnInit() {
        if (this.cellContext && this.cellContext.focusGroup) {
            this.group = this.cellContext.focusGroup;
        }
        if (this.group) {
            this.group.registerElement(this);
        }
    }
    ngAfterViewInit() {
        if (!this.element) {
            this.element = new DefaultFocusableElement(this.hostElement, this.renderer);
        }
        if (this.group) {
            const isActive = this.group.isActive;
            this.toggle(isActive);
        }
    }
    ngOnDestroy() {
        if (this.group) {
            this.group.unregisterElement(this);
        }
    }
    /**
     * @hidden
     */
    toggle(active) {
        if (this.element && active !== this.active) {
            this.active = active;
            this.element.toggle(active);
        }
    }
    /**
     * @hidden
     */
    canFocus() {
        return this.element && this.element.canFocus();
    }
    /**
     * @hidden
     */
    isNavigable() {
        return this.element && this.element.isNavigable();
    }
    /**
     * @hidden
     */
    focus() {
        if (this.element) {
            this.element.focus();
        }
    }
    /**
     * @hidden
     */
    hasFocus() {
        return this.element && this.element.hasFocus();
    }
    /**
     * @hidden
     */
    registerElement(element) {
        this.element = element;
    }
}
FocusableDirective.decorators = [
    { type: Directive, args: [{
                selector: '[kendoTreeListFocusable]' + `,
        [kendoTreeListAddCommand],
        [kendoTreeListEditCommand],
        [kendoTreeListRemoveCommand],
        [kendoTreeListSaveCommand],
        [kendoTreeListCancelCommand]
    `
            },] },
];
/** @nocollapse */
FocusableDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: FocusGroup, decorators: [{ type: Optional }] }
];
FocusableDirective.propDecorators = {
    cellContext: [{ type: Input, args: ['kendoTreeListFocusable',] }],
    addCellContext: [{ type: Input, args: ['kendoTreeListAddCommand',] }],
    editCellContext: [{ type: Input, args: ['kendoTreeListEditCommand',] }],
    removeCellContext: [{ type: Input, args: ['kendoTreeListRemoveCommand',] }],
    saveCellContext: [{ type: Input, args: ['kendoTreeListSaveCommand',] }],
    cancelCellContext: [{ type: Input, args: ['kendoTreeListCancelCommand',] }]
};
