// @progress/kendo-angular-treelist/dist/es2015/expand-state/expand-state.service.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Injectable } from '@angular/core';
import { ExpandEvent } from './expand-event';
import { Subject } from "rxjs";
/**
 * @hidden
 */
export const defaultExpanded = (_item) => true;
/**
 * @hidden
 */
export class ExpandStateService {
    constructor() {
        this.changes = new Subject();
        this.isExpanded = defaultExpanded;
    }
    toggleState(dataItem) {
        const isExpanded = this.isExpanded(dataItem);
        this.changes.next(new ExpandEvent({ dataItem: dataItem, expand: !isExpanded }));
    }
    expand(dataItem) {
        if (!this.isExpanded(dataItem)) {
            this.changes.next(new ExpandEvent({ dataItem: dataItem, expand: true }));
        }
    }
    collapse(dataItem) {
        if (this.isExpanded(dataItem)) {
            this.changes.next(new ExpandEvent({ dataItem: dataItem, expand: false }));
        }
    }
}
ExpandStateService.decorators = [
    { type: Injectable },
];
