// @progress/kendo-angular-scheduler/dist/npm/views/view-items/base-slot.service.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
/**
 * @hidden
 */
var BaseSlotService = /** @class */ (function () {
    function BaseSlotService() {
        this.containerSize = 0;
        this.slotsChange = new core_1.EventEmitter();
        this.groups = [];
    }
    BaseSlotService.prototype.registerItem = function (component) {
        var group = this.itemGroup(component);
        group.registerItem(component);
    };
    BaseSlotService.prototype.unregisterItem = function (component, id) {
        id = id || {
            index: component.index,
            resourceIndex: component.resourceIndex,
            rangeIndex: component.rangeIndex
        };
        var group = this.groups[id.resourceIndex];
        if (group) {
            group.unregisterItem(component, id);
        }
    };
    BaseSlotService.prototype.registerSlot = function (slot) {
        var group = this.slotGroup(slot);
        group.registerSlot(slot);
    };
    BaseSlotService.prototype.unregisterSlot = function (slot) {
        var group = this.groups[slot.id.resourceIndex];
        if (group) {
            group.unregisterSlot(slot);
        }
    };
    BaseSlotService.prototype.invalidate = function () {
        this.clearEmptyGroups();
        this.cleanRanges();
        this.slotsChange.emit();
        this.forEachSlot(function (slot) {
            slot.invalidate();
        });
    };
    BaseSlotService.prototype.cleanRanges = function () {
        this.groups.forEach(function (group) {
            group.cleanRanges();
        });
    };
    BaseSlotService.prototype.clearEmptyGroups = function () {
        var groups = this.groups;
        var index = this.groups.length - 1;
        while (index > 0 && !groups[index].hasSlots) {
            groups.splice(index, 1);
            index--;
        }
    };
    BaseSlotService.prototype.itemGroup = function (item) {
        var index = item.resourceIndex;
        if (!this.groups[index]) {
            this.groups[index] = this.createGroup(index);
        }
        return this.groups[index];
    };
    BaseSlotService.prototype.slotGroup = function (slot) {
        var index = slot.id.resourceIndex;
        if (!this.groups[index]) {
            this.groups[index] = this.createGroup(index);
        }
        return this.groups[index];
    };
    return BaseSlotService;
}());
exports.BaseSlotService = BaseSlotService;
