// @progress/kendo-angular-treelist/dist/es/expand-state/expand-state.service.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Injectable } from '@angular/core';
import { ExpandEvent } from './expand-event';
import { Subject } from "rxjs";
/**
 * @hidden
 */
export var defaultExpanded = function (_item) { return true; };
/**
 * @hidden
 */
var ExpandStateService = /** @class */ (function () {
    function ExpandStateService() {
        this.changes = new Subject();
        this.isExpanded = defaultExpanded;
    }
    ExpandStateService.prototype.toggleState = function (dataItem) {
        var isExpanded = this.isExpanded(dataItem);
        this.changes.next(new ExpandEvent({ dataItem: dataItem, expand: !isExpanded }));
    };
    ExpandStateService.prototype.expand = function (dataItem) {
        if (!this.isExpanded(dataItem)) {
            this.changes.next(new ExpandEvent({ dataItem: dataItem, expand: true }));
        }
    };
    ExpandStateService.prototype.collapse = function (dataItem) {
        if (this.isExpanded(dataItem)) {
            this.changes.next(new ExpandEvent({ dataItem: dataItem, expand: false }));
        }
    };
    ExpandStateService.decorators = [
        { type: Injectable },
    ];
    return ExpandStateService;
}());
export { ExpandStateService };
