// @progress/kendo-angular-treelist/dist/npm/expand-state/expandable.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var treelist_component_1 = require("../treelist.component");
var kendo_common_1 = require("@progress/kendo-common");
var utils_1 = require("../utils");
/**
 * A directive which controls the expanded state of the items
 */
var ExpandableDirective = /** @class */ (function () {
    function ExpandableDirective(treelist) {
        this.treelist = treelist;
        /**
         * Fires when the expandedKeys are changed.
         */
        this.expandedKeysChange = new core_1.EventEmitter();
        /**
         * Specifies if the items should be initially expanded.
         * @default false
         */
        this.initiallyExpanded = false;
        this.state = new Set();
        this.treelist.isExpanded = this.isExpanded.bind(this);
        this.toggleState = this.toggleState.bind(this);
        this.subscriptions = this.treelist.expandStateChange.subscribe(this.toggleState);
    }
    Object.defineProperty(ExpandableDirective.prototype, "expandedKeys", {
        get: function () {
            return Array.from(this.state.keys());
        },
        /**
         * Sets the expanded item keys.
         */
        set: function (value) {
            if (value !== this._expandedKeys) {
                this._expandedKeys = value;
                this.treelist.updateView();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExpandableDirective.prototype, "expandBy", {
        /**
         * Defines the item key that will be stored in the expandedKeys collection.
         * Defaults to the TreeList idField.
         */
        set: function (key) {
            if (utils_1.isString(key)) {
                this._expandBy = kendo_common_1.getter(key);
            }
            else {
                this._expandBy = key;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExpandableDirective.prototype, "keyGetter", {
        get: function () {
            return this._expandBy || this.treelist.idGetter;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @hidden
     */
    ExpandableDirective.prototype.isExpanded = function (item) {
        var key = this.keyGetter(item);
        return this.state.has(key) ? !this.initiallyExpanded : this.initiallyExpanded;
    };
    ExpandableDirective.prototype.ngOnChanges = function (changes) {
        if (changes.expandedKeys) {
            this.state = new Set(this._expandedKeys);
        }
        else if (changes.initiallyExpanded) {
            this.state.clear();
        }
    };
    ExpandableDirective.prototype.ngOnDestroy = function () {
        this.subscriptions.unsubscribe();
    };
    ExpandableDirective.prototype.toggleState = function (args) {
        var key = this.keyGetter(args.dataItem);
        if (Boolean(this.initiallyExpanded) !== args.expand) {
            this.state.add(key);
        }
        else {
            this.state.delete(key);
        }
        this.expandedKeysChange.emit(this.expandedKeys);
    };
    ExpandableDirective.decorators = [
        { type: core_1.Directive, args: [{
                    exportAs: 'kendoTreeListExpandable',
                    selector: '[kendoTreeListExpandable]'
                },] },
    ];
    /** @nocollapse */
    ExpandableDirective.ctorParameters = function () { return [
        { type: treelist_component_1.TreeListComponent }
    ]; };
    ExpandableDirective.propDecorators = {
        expandedKeysChange: [{ type: core_1.Output }],
        expandedKeys: [{ type: core_1.Input }],
        initiallyExpanded: [{ type: core_1.Input }],
        expandBy: [{ type: core_1.Input }]
    };
    return ExpandableDirective;
}());
exports.ExpandableDirective = ExpandableDirective;
