// @progress/kendo-angular-treelist/dist/npm/expand-state/expand-state.service.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var expand_event_1 = require("./expand-event");
var rxjs_1 = require("rxjs");
/**
 * @hidden
 */
exports.defaultExpanded = function (_item) { return true; };
/**
 * @hidden
 */
var ExpandStateService = /** @class */ (function () {
    function ExpandStateService() {
        this.changes = new rxjs_1.Subject();
        this.isExpanded = exports.defaultExpanded;
    }
    ExpandStateService.prototype.toggleState = function (dataItem) {
        var isExpanded = this.isExpanded(dataItem);
        this.changes.next(new expand_event_1.ExpandEvent({ dataItem: dataItem, expand: !isExpanded }));
    };
    ExpandStateService.prototype.expand = function (dataItem) {
        if (!this.isExpanded(dataItem)) {
            this.changes.next(new expand_event_1.ExpandEvent({ dataItem: dataItem, expand: true }));
        }
    };
    ExpandStateService.prototype.collapse = function (dataItem) {
        if (this.isExpanded(dataItem)) {
            this.changes.next(new expand_event_1.ExpandEvent({ dataItem: dataItem, expand: false }));
        }
    };
    ExpandStateService.decorators = [
        { type: core_1.Injectable },
    ];
    return ExpandStateService;
}());
exports.ExpandStateService = ExpandStateService;
