// @progress/kendo-angular-treeview/dist/es/drag-and-drop/drag-clue/drag-clue.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Component, HostBinding, ChangeDetectorRef, ChangeDetectionStrategy } from "@angular/core";
import { DropAction } from '../models';
/**
 * @hidden
 */
var DragClueComponent = /** @class */ (function () {
    function DragClueComponent(cdr) {
        this.cdr = cdr;
        this.hostClasses = true;
        this.posistionStyle = 'fixed';
    }
    Object.defineProperty(DragClueComponent.prototype, "statusIconClass", {
        get: function () {
            switch (this.action) {
                case DropAction.Add: return 'k-i-plus';
                case DropAction.InsertTop: return 'k-i-insert-up';
                case DropAction.InsertBottom: return 'k-i-insert-down';
                case DropAction.InsertMiddle: return 'k-i-insert-middle';
                case DropAction.Invalid:
                default: return 'k-i-cancel';
            }
        },
        enumerable: true,
        configurable: true
    });
    // exposed as a public method that can be called from outside as the component uses `OnPush` strategy
    DragClueComponent.prototype.detectChanges = function () {
        this.cdr.detectChanges();
    };
    DragClueComponent.decorators = [
        { type: Component, args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-treeview-drag-clue',
                    template: "\n        <ng-container *ngIf=\"!template\">\n            <span class=\"k-icon {{statusIconClass}} k-drag-status\"></span>\n            <span>{{text}}</span>\n        </ng-container>\n\n        <ng-template\n            *ngIf=\"template\"\n            [ngTemplateOutlet]=\"template\"\n            [ngTemplateOutletContext]=\"{\n                text: text,\n                action: action,\n                sourceItem: sourceItem,\n                destinationItem: destinationItem\n            }\"\n        >\n        </ng-template>\n    "
                },] },
    ];
    /** @nocollapse */
    DragClueComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    DragClueComponent.propDecorators = {
        hostClasses: [{ type: HostBinding, args: ['class.k-header',] }, { type: HostBinding, args: ['class.k-drag-clue',] }],
        posistionStyle: [{ type: HostBinding, args: ['style.position',] }]
    };
    return DragClueComponent;
}());
export { DragClueComponent };
