// @progress/kendo-angular-treelist/dist/npm/navigation/focusable.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var default_focusable_element_1 = require("./default-focusable-element");
var focus_group_1 = require("./focus-group");
/**
 * A directive that controls the way focusable cell elements receive
 * [focus in a navigable TreeList]({% slug keyboard_navigation_treelist %}).
 *
 * The directive takes as input the `cellContext` from the cell template.
 *
 * ```ts-no-run
 * <kendo-treelist-column>
 *   <ng-template kendoTreeListCellTemplate let-cellContext="cellContext">
 *     <input type="button" class="k-button" [kendoTreeListFocusable]="cellContext" value="Focus me!" />
 *   </ng-template>
 * </kendo-treelist-column>
 * ```
 */
var FocusableDirective = /** @class */ (function () {
    function FocusableDirective(hostElement, renderer, group) {
        this.hostElement = hostElement;
        this.renderer = renderer;
        this.group = group;
        this.active = true;
    }
    Object.defineProperty(FocusableDirective.prototype, "addCellContext", {
        /**
         * @hidden
         */
        set: function (value) { this.cellContext = value; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FocusableDirective.prototype, "editCellContext", {
        /**
         * @hidden
         */
        set: function (value) { this.cellContext = value; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FocusableDirective.prototype, "removeCellContext", {
        /**
         * @hidden
         */
        set: function (value) { this.cellContext = value; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FocusableDirective.prototype, "saveCellContext", {
        /**
         * @hidden
         */
        set: function (value) { this.cellContext = value; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FocusableDirective.prototype, "cancelCellContext", {
        /**
         * @hidden
         */
        set: function (value) { this.cellContext = value; },
        enumerable: true,
        configurable: true
    });
    FocusableDirective.prototype.ngOnInit = function () {
        if (this.cellContext && this.cellContext.focusGroup) {
            this.group = this.cellContext.focusGroup;
        }
        if (this.group) {
            this.group.registerElement(this);
        }
    };
    FocusableDirective.prototype.ngAfterViewInit = function () {
        if (!this.element) {
            this.element = new default_focusable_element_1.DefaultFocusableElement(this.hostElement, this.renderer);
        }
        if (this.group) {
            var isActive = this.group.isActive;
            this.toggle(isActive);
        }
    };
    FocusableDirective.prototype.ngOnDestroy = function () {
        if (this.group) {
            this.group.unregisterElement(this);
        }
    };
    /**
     * @hidden
     */
    FocusableDirective.prototype.toggle = function (active) {
        if (this.element && active !== this.active) {
            this.active = active;
            this.element.toggle(active);
        }
    };
    /**
     * @hidden
     */
    FocusableDirective.prototype.canFocus = function () {
        return this.element && this.element.canFocus();
    };
    /**
     * @hidden
     */
    FocusableDirective.prototype.isNavigable = function () {
        return this.element && this.element.isNavigable();
    };
    /**
     * @hidden
     */
    FocusableDirective.prototype.focus = function () {
        if (this.element) {
            this.element.focus();
        }
    };
    /**
     * @hidden
     */
    FocusableDirective.prototype.hasFocus = function () {
        return this.element && this.element.hasFocus();
    };
    /**
     * @hidden
     */
    FocusableDirective.prototype.registerElement = function (element) {
        this.element = element;
    };
    FocusableDirective.decorators = [
        { type: core_1.Directive, args: [{
                    selector: '[kendoTreeListFocusable]' + ",\n        [kendoTreeListAddCommand],\n        [kendoTreeListEditCommand],\n        [kendoTreeListRemoveCommand],\n        [kendoTreeListSaveCommand],\n        [kendoTreeListCancelCommand]\n    "
                },] },
    ];
    /** @nocollapse */
    FocusableDirective.ctorParameters = function () { return [
        { type: core_1.ElementRef },
        { type: core_1.Renderer2 },
        { type: focus_group_1.FocusGroup, decorators: [{ type: core_1.Optional }] }
    ]; };
    FocusableDirective.propDecorators = {
        cellContext: [{ type: core_1.Input, args: ['kendoTreeListFocusable',] }],
        addCellContext: [{ type: core_1.Input, args: ['kendoTreeListAddCommand',] }],
        editCellContext: [{ type: core_1.Input, args: ['kendoTreeListEditCommand',] }],
        removeCellContext: [{ type: core_1.Input, args: ['kendoTreeListRemoveCommand',] }],
        saveCellContext: [{ type: core_1.Input, args: ['kendoTreeListSaveCommand',] }],
        cancelCellContext: [{ type: core_1.Input, args: ['kendoTreeListCancelCommand',] }]
    };
    return FocusableDirective;
}());
exports.FocusableDirective = FocusableDirective;
