// @progress/kendo-angular-treelist/dist/es2015/filtering/cell/filter-cell.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Component, Input } from '@angular/core';
import { ColumnComponent } from '../../columns/column.component';
import { cloneFilters } from '../../common/filter-descriptor-differ';
import { FocusGroup } from '../../navigation/focus-group';
import { isNullOrEmptyString, isPresent } from '../../utils';
/**
 * @hidden
 */
export class FilterCellComponent {
    constructor(focusGroup) {
        this.focusGroup = focusGroup;
        this._templateContext = {};
    }
    get filter() {
        return this._filter;
    }
    set filter(value) {
        this._filter = cloneFilters(value);
    }
    get templateContext() {
        this._templateContext.$implicit = this.filter;
        this._templateContext.column = this.column;
        this._templateContext.filter = this.filter;
        this._templateContext.cellContext = { focusGroup: this.focusGroup };
        return this._templateContext;
    }
    get hasTemplate() {
        return isPresent(this.column.filterCellTemplateRef);
    }
    get isFilterable() {
        return isPresent(this.column) && !isNullOrEmptyString(this.column.field) && this.column.filterable;
    }
}
FilterCellComponent.decorators = [
    { type: Component, args: [{
                selector: '[kendoTreeListFilterCell]',
                template: `
        <ng-template [ngIf]="isFilterable">
            <ng-container [ngSwitch]="hasTemplate">
                <ng-container *ngSwitchCase="false">
                    <ng-container kendoFilterCellHost [column]="column" [filter]="filter"></ng-container>
                </ng-container>
                <ng-container *ngSwitchCase="true">
                    <ng-template
                        *ngIf="column.filterCellTemplateRef"
                        [ngTemplateOutlet]="column.filterCellTemplateRef"
                        [ngTemplateOutletContext]="templateContext">
                    </ng-template>
                </ng-container>
            </ng-container>
        </ng-template>
    `
            },] },
];
/** @nocollapse */
FilterCellComponent.ctorParameters = () => [
    { type: FocusGroup }
];
FilterCellComponent.propDecorators = {
    column: [{ type: Input }],
    filter: [{ type: Input }]
};
