// @progress/kendo-angular-charts/dist/es2015/chart/series-item/highlight.component.generated.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Input } from '@angular/core';
import { SettingsComponent } from '../../common/settings.component';
/**
 * @hidden
 */
export class SeriesHighlightComponentGenerated extends SettingsComponent {
    constructor(configurationService) {
        super('highlight', configurationService);
        this.configurationService = configurationService;
    }
}
SeriesHighlightComponentGenerated.propDecorators = {
    border: [{ type: Input }],
    color: [{ type: Input }],
    line: [{ type: Input }],
    markers: [{ type: Input }],
    opacity: [{ type: Input }],
    toggle: [{ type: Input }],
    visible: [{ type: Input }],
    visual: [{ type: Input }]
};
