// @progress/kendo-angular-scheduler/dist/es/views/common/hint-container.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Component, ChangeDetectorRef, ContentChild, TemplateRef } from '@angular/core';
/**
 * @hidden
 */
var HintContainerComponent = /** @class */ (function () {
    function HintContainerComponent(changeDetector) {
        this.changeDetector = changeDetector;
        this.changeDetector.detach();
    }
    HintContainerComponent.prototype.detectChanges = function () {
        this.changeDetector.detectChanges();
    };
    HintContainerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'kendo-hint-container',
                    template: "\n        <ng-container [ngTemplateOutlet]=\"hintTemplate\">\n        </ng-container>\n    "
                },] },
    ];
    /** @nocollapse */
    HintContainerComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    HintContainerComponent.propDecorators = {
        hintTemplate: [{ type: ContentChild, args: [TemplateRef,] }]
    };
    return HintContainerComponent;
}());
export { HintContainerComponent };
